src/main/java/de/tum/in/www1/artemis/repository/CompetencyRepository.java

```diff
@@ -47,17 +50,19 @@ public interface CompetencyRepository extends JpaRepository<Competency, Long> {
     Optional<Competency> findByIdWithLectureUnits(@Param("competencyId") long competencyId);
 
     @Query("""
-            SELECT c
-            FROM Competency c
-                LEFT JOIN FETCH c.userProgress
-                LEFT JOIN FETCH c.exercises
-                LEFT JOIN FETCH c.lectureUnits lu
-                LEFT JOIN FETCH lu.completedUsers
-                LEFT JOIN FETCH lu.lecture
-                LEFT JOIN FETCH lu.exercise
-            WHERE c.id = :competencyId
+            SELECT competency
+            FROM Competency competency
+                LEFT JOIN competency.userProgress progress
+                    ON competency.id = progress.learningGoal.id AND progress.user.id = :userId
+                LEFT JOIN FETCH competency.exercises
+                LEFT JOIN FETCH competency.lectureUnits lectureUnits
+                LEFT JOIN lectureUnits.completedUsers completedUsers
+                    ON lectureUnits.id = completedUsers.lectureUnit.id AND completedUsers.user.id = :userId
+                LEFT JOIN FETCH lectureUnits.lecture
+            WHERE competency.id = :competencyId
             """)
-    Optional<Competency> findByIdWithExercisesAndLectureUnits(@Param("competencyId") Long competencyId);
+    @EntityGraph(type = LOAD, attributePaths = { "userProgress", "lectureUnits.completedUsers" })
```

Comment:
is this `EntityGraph` really needed? I think we want to avoid loading all `lectureUnits.completedUsers`
Can you not simply use left join fetch as above but add a WHERE clause?

Reply:
Fetching all data is prevented by joining on `userId`. Adding this to the WHERE clause would incorrectly return the competency only when a progress entity for the user exists. If not, the query would not find it and return empty.

Reply:
Can you use `LEFT JOIN FETCH ... ON ` in the query instead of the entity graph?

Reply:
This is what I tried at first. It's not supported.

Reply:
can you add a comment above? I think the statement is not fully clear otherwise.

Reply:
--> follow-up